### ISSUE TITLE
SMP and IPI for ESP32-S3

### ISSUE THREAD

**Issue by zap8600** [jcmvbkbc/linux-xtensa#6]
>> IPI trigger is not a part of the esp32 interrupt matrix and it doesn't have to be a part of an interrupt controller in the device tree, it may be a standalone device.

> So we need to find this device?

By 'device' I meant 'a node in the device tree'. The hardware itself is the system registers `SYSTEM_CPU_INTR_FROM_CPU_*_REG`.

I'd also suggest discussing it in a separate issue, as there's a lot of off-topic here already.

_Originally posted by @jcmvbkbc in https://github.com/jcmvbkbc/linux-xtensa/issues/5#issuecomment-1618998129_

**Comment by zap8600** [comment]
So how can we find the node to add to the device tree? Are there any ESP-IDF functions we can use to find it?

**Comment by jcmvbkbc** [comment]
> So how can we find the node to add to the device tree?

I'm not sure I understand the question. We need to just literally add a node to the esp32s3.dts. It should have the `compatible` property that would match a device driver that must be written to control it and the `reg` property that would point to the right registers.

**Comment by zap8600** [comment]
> I'm not sure I understand the question. We need to just literally add a node to the esp32s3.dts. It should have the `compatible` property that would match a device driver that must be written to control it and the `reg` property that would point to the right registers.

What sort of node are we looking for? Can we use ESP-IDF functions to find the node and its details? Is there a DTS in the ESP-IDF we could look at? Where in the DTS would we put the node?

**Comment by jcmvbkbc** [comment]
> Can we use ESP-IDF functions to find the node and its details?

I don't think so.

> Is there a DTS in the ESP-IDF we could look at? 

No.

> What sort of node are we looking for? Where in the DTS would we put the node?

I'd suggest starting with the [specification](https://www.devicetree.org/specifications/) and then looking at other dts files in the linux kernel tree.

**Comment by zap8600** [comment]
> I'd suggest starting with the [specification](https://www.devicetree.org/specifications/) and then looking at other dts files in the linux kernel tree.

Here is what I think is right so far.
```
cpu_0 {
    reg = <0x600C0000 0x30>;
};
```
I was unsure what to name it. I believe this would go under `soc` in the DTS. I still need to find what `compatible` should be. Anything else I need to do?

**Comment by jcmvbkbc** [comment]
> reg = <0x600C0000 0x30>

That covers some unrelated registers and doesn't cover any IPI registers. Perhaps it should be split into multiple regions.

> I was unsure what to name it. 

Normally it should be `name@base_address`, e.g.: `ipi@600c0000`.

> I believe this would go under soc in the DTS.

Yes, that sounds right.

> I still need to find what compatible should be

It needs to be specific enough to indicate what hardware it is for, but otherwise it's up to you.

> Anything else I need to do?

The driver that would service this device and interact with the SMP guts of the xtensa linux kernel.

**Comment by zap8600** [comment]
> That covers some unrelated registers and doesn't cover any IPI registers. Perhaps it should be split into multiple regions.

My bad. I thought it was what I should put for `SYSTEM_CPU_INTR_FROM_CPU_0_REG`.
> It needs to be specific enough to indicate what hardware it is for, but otherwise it's up to you.

Would something like `compatible = "esp,esp32s3-ipi";` work?

**Comment by jcmvbkbc** [comment]
> I thought it was what I should put for SYSTEM_CPU_INTR_FROM_CPU_0_REG.

The format of `reg` property is `<base_address_cells size_cells …>`, with more base/size pairs in the place of … if necessary to represent multiple regions. E.g. to cover the `SYSTEM_CPU_INTR_FROM_CPU_0_REG` alone it would be `<0x600c0030 4>`

> Would something like compatible = "esp,esp32s3-ipi"; work?

Yes.

**Comment by zap8600** [comment]
> The format of `reg` property is `<base_address_cells size_cells …>`, with more base/size pairs in the place of … if necessary to represent multiple regions. E.g. to cover the `SYSTEM_CPU_INTR_FROM_CPU_0_REG` alone it would be `<0x600c0030 4>`

So would this work?
```
ipi: ipi@600c0000 {
    compatible = "esp,esp32s3-ipi";
    cpu_0 {
        reg = <0x34 0x38>;
    };
    cpu_1 {
        reg = <0x38 0x3c>;
    };
};
```

**Comment by jcmvbkbc** [comment]
> So would this work?

Nope. It has wrong sizes and it implies certain address translation which is not set up here. Please read the DTS spec for the details.
Again, to cover the `SYSTEM_CPU_INTR_FROM_CPU_0_REG` alone the reg would be `<0x600c0030 4>`.

**Comment by zap8600** [comment]
> Nope. It has wrong sizes and it implies certain address translation which is not set up here. Please read the DTS spec for the details. Again, to cover the `SYSTEM_CPU_INTR_FROM_CPU_0_REG` alone the reg would be `<0x600c0030 4>`.

How about this?
```
ipi: ipi@600c0000 {
    compatible = "esp,esp32s3-ipi";
    cpu0@30 {
        reg = <0x0 0x4>;
    };
    cpu1@34 {
        reg = <0x34 0x4>;
    };
};

**Comment by jcmvbkbc** [comment]
> How about this?

How about just building your kernel with this dts and checking how it works?

**Comment by zap8600** [comment]
> How about just building your kernel with this dts and checking how it works?

Do I need to add another `cpu` to `cpus` first?

**Comment by jcmvbkbc** [comment]
> Do I need to add another cpu to cpus first?

It doesn't matter: there's no code that would interpret this information.

**Comment by zap8600** [comment]
> It doesn't matter: there's no code that would interpret this information.

Alright. Will the build fail if my DTS is invalid?

**Comment by jcmvbkbc** [comment]
> Will the build fail if my DTS is invalid?

Depends on how invalid. Syntactically invalid dts will break the build, yes. Semantically inconsistent may cause dtc to print warnings. Otherwise you'll see issues at runtime.

**Comment by zap8600** [comment]
I've been messing with my PC for hours trying to get it to build the base kernel. I'll leave my kernel and rootfs here in case anyone needs it. Now I have to build the kernel with my DTS.
[esp32s3linux.zip](https://github.com/jcmvbkbc/linux-xtensa/files/11982219/esp32s3linux.zip)

**Comment by zap8600** [comment]
I'm going to delay this for now. I'm waiting on some hardware to arrive. I have an off-topic question though. Would it be possible to wire an SD Card to the board (over either SDIO or SPI), and then give Linux access to it? I just want to know if it's possible. If it is then I'll look into implementing it later.

**Comment by jcmvbkbc** [comment]
> Would it be possible to wire an SD Card to the board (over either SDIO or SPI), and then give Linux access to it?

It should be. I don't see why it shouldn't be possible.

**Comment by raspiduino** [comment]
> > Would it be possible to wire an SD Card to the board (over either SDIO or SPI), and then give Linux access to it?
> 
> It should be. I don't see why it shouldn't be possible.

Then it should be perfect for esp32 cam module

**Comment by zap8600** [comment]
I'm going to close this for now. I think I should start with a less complex part of the chip to implement, like GPIO or SPI. I'll create an issue for questions on how I should implement these once I get to it.
